Firestore/core/src/remote/grpc_adapt/grpc_swift_byte_buffer.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FIRESTORE_CORE_SRC_REMOTE_GRPC_ADAPT_SLICE_H_
+#define FIRESTORE_CORE_SRC_REMOTE_GRPC_ADAPT_SLICE_H_
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "Firestore/core/src/remote/grpc_adapt/grpc_swift_channel.h"
+#include "Firestore/core/src/remote/grpc_adapt/grpc_swift_status.h"
+#include "Firestore/core/src/remote/grpc_adapt/grpc_swift_string_ref.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+namespace grpc_adapt {
+
+class ByteBufferShim;
+class Slice;
+
+class ByteBuffer final {
+ public:
+  /// Constuct an empty buffer.
+  ByteBuffer();
+  ByteBuffer(const Slice* slices, size_t nslices);
+
+  /// Buffer size in bytes.
+  size_t Length() const;
+  /// Dump (read) the buffer contents into \a slices.
+  Status Dump(std::vector<Slice>* slices) const;
+
+ private:
+  ByteBufferShim* shim_;
```

Comment:
Headers cannot reference to Swift shims..because they will be included in .cc files.

You can have a ByteBufferImpl class defined in .mm file, and subclass ByteBuffer.

Reply:
Can I also not use forward declaration?

Reply:
FWIW, you can probably follow `grpc_swift_channel.h/mm`. There is a `ClientContext` that does similar things.